userspace/libsinsp/container.cpp

```diff
@@ -551,53 +552,66 @@ void sinsp_container_manager::create_engines() {
 		                                                                   m_static_name,
 		                                                                   m_static_image);
 		m_container_engines.push_back(engine);
-		m_container_engine_by_type[CT_STATIC] = engine;
+		m_container_engine_by_type[CT_STATIC].push_back(engine);
 		return;
 	}
 #if !defined(MINIMAL_BUILD) && !defined(__EMSCRIPTEN__)
 #ifndef _WIN32
 	if(m_container_engine_mask & (1 << CT_PODMAN)) {
 		auto podman_engine = std::make_shared<container_engine::podman>(*this);
 		m_container_engines.push_back(podman_engine);
-		m_container_engine_by_type[CT_PODMAN] = podman_engine;
+		m_container_engine_by_type[CT_PODMAN].push_back(podman_engine);
 	}
 	if(m_container_engine_mask & (1 << CT_DOCKER)) {
 		auto docker_engine = std::make_shared<container_engine::docker_linux>(*this);
 		m_container_engines.push_back(docker_engine);
-		m_container_engine_by_type[CT_DOCKER] = docker_engine;
+		m_container_engine_by_type[CT_DOCKER].push_back(docker_engine);
 	}
 
 	if(m_container_engine_mask & ((1 << CT_CRI) | (1 << CT_CRIO) | (1 << CT_CONTAINERD))) {
-		auto cri_engine = std::make_shared<container_engine::cri>(*this);
-		m_container_engines.push_back(cri_engine);
-		m_container_engine_by_type[CT_CRI] = cri_engine;
-		m_container_engine_by_type[CT_CRIO] = cri_engine;
-		m_container_engine_by_type[CT_CONTAINERD] = cri_engine;
+		// Get CRI socket paths from settings
+		libsinsp::cri::cri_settings& cri_settings = libsinsp::cri::cri_settings::get();
+		if(cri_settings.get_cri_unix_socket_paths().empty()) {
+			// Add default paths
+			cri_settings.add_cri_unix_socket_path("/run/containerd/containerd.sock");
+			cri_settings.add_cri_unix_socket_path("/run/crio/crio.sock");
+			cri_settings.add_cri_unix_socket_path("/run/k3s/containerd/containerd.sock");
+		}
+
+		const auto& cri_socket_paths = cri_settings.get_cri_unix_socket_paths();
```

Comment:
Can we move default values inside here?
I mean, it would be easier if the consumer doesn't need to know about the defaults, and instead just calls this API that, when vector is empty, pushes the default values (that at that point can be a function local static vector)

Reply:
Theoretically, I agree with you. I opted for this choice to emulate the [old behavior](https://github.com/falcosecurity/libs/pull/2141/files/fded35bbfab8f504c71cde75fe9ad44033753baa#diff-4d251d1ba12fc4efc684aca05049f7b29107725c527047758fe0a6fc8149763cL58). If I had used lazy loading default inside `cri_settings.get_cri_unix_socket_paths()`, `cri_settings.get_cri_unix_socket_paths().empty()` would be always false, slightly changing the behavior from a consumer perspective.

Plus, I assume this implementation will soon be replaced by https://github.com/falcosecurity/falco/issues/3403, so it makes sense to me to introduce as few changes as possible (thus, I avoided possible API enhancements as well).

Other than that, I don't have strong opinions. So, if my reasoning does not convince you, let me know, and I will apply your suggestion.

Thank you :pray:

Reply:
That's ok! no problem then!